File: go/sampling.go
// ADR-097 — session-scoped sampling. Go port of packages/core/src/sampling.ts.
//
// THE CORE RULE: the sampling decision is made ONCE per session (or per trace,
// where one exists) and applies to EVERY log line under it. The invariant this
// buys: any trace you can open has 100% of its log lines. Never a partial view.
//
// Every vector this file must reproduce lives in parity/sampling-corpus.json
// and is asserted by parity_corpus_test.go — the same file the TS, Rust, Python
// and .NET lanes load. See parity/README.md for the hash derivation and the
// porting traps.

package observability

import (
	"math"
	"strings"
)

const (
	fnvOffsetBasis32 uint32 = 0x811c9dc5
	fnvPrime32       uint32 = 0x01000193
	// twoPow32 as an exact float64 — dividing a uint32 by this is a pure
	// exponent adjustment, so every language gets the identical double.
	twoPow32 float64 = 4294967296.0
)

// FNV1a32 computes FNV-1a 32-bit over the UTF-8 bytes of s.
//
// XOR-then-multiply (this is FNV-*1a*, not FNV-1). Go's uint32 arithmetic wraps
// natively; ranging over a string with an index yields bytes, not runes, which
// is what the byte-at-a-time fold requires.
func FNV1a32(s string) uint32 {
	h := fnvOffsetBasis32
	for i := 0; i < len(s); i++ {
		h ^= uint32(s[i])
		h *= fnvPrime32
	}
	return h
}

// SampleDecision is the one sampling primitive. Deterministic and stable for
// the lifetime of an id.
//
//   - non-finite ratio -> IN (fail open — telemetry going dark on a config
//     hiccup is the failure ADR-097 forbids; x < NaN is false everywhere, so the
//     naive path would sample *everything* out)
//   - ratio <= 0.0 -> OUT, ratio >= 1.0 -> IN, both taken before any float math
//     so 1.0 can never drop and 0.0 can never keep
//   - otherwise (hash / 2^32) < ratio, strict less-than
func SampleDecision(id string, ratio float64) bool {
	if math.IsNaN(ratio) || math.IsInf(ratio, 0) {
		return true
	}
	if ratio <= 0.0 {
		return false
	}
	if ratio >= 1.0 {
		return true
	}
	// float64(uint32) is exact — the uint32 must NOT be routed through int32,
	// which would make hashes above 2^31 negative and flip their decision.
	return float64(FNV1a32(id))/twoPow32 < ratio
}

// CanonicalLevel is an ADR-096 canonical log level, uppercase.
//
// Distinct from Level, which is the lowercase *event wire* severity on the
// ingest envelope. This is the log-line severity the ClickHouse queries filter
// on, and `level IN ('ERROR','FATAL')` is CASE-SENSITIVE — emitting "error"
// silently makes every error a non-error.
type CanonicalLevel string

const (
	CanonicalTrace CanonicalLevel = "TRACE"
	CanonicalDebug CanonicalLevel = "DEBUG"
	CanonicalInfo  CanonicalLevel = "INFO"
	CanonicalWarn  CanonicalLevel = "WARN"
	CanonicalError CanonicalLevel = "ERROR"
	CanonicalFatal CanonicalLevel = "FATAL"
)

// levelRank is the ordering used by the minimum-level filter. Matches OTel
// severity numbers.
var levelRank = map[CanonicalLevel]int{
	CanonicalTrace: 1,
	CanonicalDebug: 5,
	CanonicalInfo:  9,
	CanonicalWarn:  13,
	CanonicalError: 17,
	CanonicalFatal: 21,
}

// levelAliases are the spellings every SDK must accept. Part of the parity
// contract.
var levelAliases = map[string]CanonicalLevel{
	"trace":       CanonicalTrace,
	"verbose":     CanonicalTrace,
	"debug":       CanonicalDebug,
	"info":        CanonicalInfo,
	"information": CanonicalInfo,
	"log":         CanonicalInfo,
	"notice":      CanonicalInfo,
	"warn":        CanonicalWarn,
	"warning":     CanonicalWarn,
	"error":       CanonicalError,
	"err":         CanonicalError,
	"fatal":       CanonicalFatal,
	"critical":    CanonicalFatal,
	"crit":        CanonicalFatal,
	"emergency":   CanonicalFatal,
	"panic":       CanonicalFatal,
}

// ParseLevel is a strict parse: a known level spelling, or ok=false.
func ParseLevel(level string) (CanonicalLevel, bool) {
	l, ok := levelAliases[strings.ToLower(strings.TrimSpace(level))]
	return l, ok
}

// NormalizeLevel normalizes any level spelling to the canonical form.
//
// Unknown spellings become INFO — fail-safe: an unrecognised level must never
// cause a drop, and must never be promoted into ERROR (which would corrupt the
// error rate).
func NormalizeLevel(level string) CanonicalLevel {
	if l, ok := ParseLevel(level); ok {
		return l
	}
	return CanonicalInfo
}

// MeetsMinimumLevel reports whether level is at or above minimum.
func MeetsMinimumLevel(level, minimum CanonicalLevel) bool {
	return levelRank[level] >= levelRank[minimum]
}

// LogSamplingInput is the input to ShouldEmitLog.
type LogSamplingInput struct {
	// Level as emitted by the caller; normalized internally.
	Level string
	// SessionID is the stable per-page session id, used when no trace exists.
	SessionID string
	// TraceSampled is the trace's own sampling decision, when a trace context
	// exists (nil when it does not). Where a trace exists its decision WINS, so
	// spans and logs never disagree.
	TraceSampled *bool
	// Enabled is the kill switch — false disables all emission, errors included.
	Enabled bool
	// MinimumLevel is the floor to emit at.
	MinimumLevel CanonicalLevel
	// LogSamplingRatio is the session-scoped browser log sampling ratio.
	LogSamplingRatio float64
}

// ShouldEmitLog is the single decision point for "does this log line get
// emitted?".
//
// Order matters and is part of the parity contract:
//  1. kill switch — off means off, no exceptions
//  2. minimum level — below the floor is not emitted
//  3. WARN/ERROR/FATAL — always 100% (ADR-010: "sampling errors is malpractice")
//  4. trace decision, if a trace exists — inherited, never re-rolled
//  5. otherwise the session decision — one roll for the whole session
func ShouldEmitLog(in LogSamplingInput) bool {
	if !in.Enabled {
		return false
	}
	level := NormalizeLevel(in.Level)
	if !MeetsMinimumLevel(level, in.MinimumLevel) {
		return false
	}
	if levelRank[level] >= levelRank[CanonicalWarn] {
		return true
	}
	if in.TraceSampled != nil {
		return *in.TraceSampled
	}
	return SampleDecision(in.SessionID, in.LogSamplingRatio)
}

File: dotnet/src/SmooAI.Observability/Sampling.cs
namespace SmooAI.Observability;

/// <summary>
/// ADR-096 canonical log level, uppercase.
/// </summary>
/// <remarks>
/// Distinct from <see cref="Level"/>, which is the lowercase <em>event wire</em>
/// severity on the ingest envelope. This is the log-line severity the ClickHouse
/// queries filter on, and <c>level IN ('ERROR','FATAL')</c> is CASE-SENSITIVE —
/// emitting <c>"error"</c> silently makes every error a non-error.
/// </remarks>
public enum CanonicalLevel
{
    Trace,
    Debug,
    Info,
    Warn,
    Error,
    Fatal,
}

/// <summary>Input to <see cref="Sampling.ShouldEmitLog"/>.</summary>
/// <param name="Level">Level as emitted by the caller; normalized internally.</param>
/// <param name="SessionId">Stable per-page session id, used when no trace exists.</param>
/// <param name="Enabled">Kill switch — false disables all emission, errors included.</param>
/// <param name="MinimumLevel">Minimum level to emit.</param>
/// <param name="LogSamplingRatio">Session-scoped browser log sampling ratio.</param>
/// <param name="TraceSampled">
/// The trace's own sampling decision, when a trace context exists (null when it
/// does not). Where a trace exists its decision WINS, so spans and logs never
/// disagree about whether a request was recorded.
/// </param>
public readonly record struct LogSamplingInput(
    string Level,
    string SessionId,
    bool Enabled,
    CanonicalLevel MinimumLevel,
    double LogSamplingRatio,
    bool? TraceSampled = null);

/// <summary>
/// ADR-097 — session-scoped sampling. .NET port of
/// <c>packages/core/src/sampling.ts</c>.
/// </summary>
/// <remarks>
/// <para>
/// THE CORE RULE: the sampling decision is made ONCE per session (or per trace,
/// where one exists) and applies to EVERY log line under it. The invariant this
/// buys: any trace you can open has 100% of its log lines. Never a partial view.
/// </para>
/// <para>
/// Every vector this class must reproduce lives in
/// <c>parity/sampling-corpus.json</c> and is asserted by
/// <c>ParityCorpusTests</c> — the same file the TS, Rust, Python and Go lanes
/// load. See <c>parity/README.md</c> for the hash derivation and the porting
/// traps.
/// </para>
/// </remarks>
public static class Sampling
{
    private const uint FnvOffsetBasis32 = 0x811c9dc5;
    private const uint FnvPrime32 = 0x01000193;

    // 2^32 as an exact double — dividing a uint by this is a pure exponent
    // adjustment, so every language gets the identical double.
    private const double TwoPow32 = 4294967296.0;

    /// <summary>
    /// FNV-1a 32-bit over the UTF-8 bytes of <paramref name="value"/>.
    /// </summary>
    /// <remarks>
    /// XOR-then-multiply (this is FNV-<em>1a</em>, not FNV-1). <c>uint</c>
    /// arithmetic is unchecked and wraps natively; the result must NOT be routed
    /// through <c>int</c>, which would make hashes above 2^31 negative and flip
    /// their sampling decision.
    /// </remarks>
    public static uint Fnv1a32(string value)
    {
        ArgumentNullException.ThrowIfNull(value);

        var h = FnvOffsetBasis32;
        foreach (var b in System.Text.Encoding.UTF8.GetBytes(value))
        {
            h ^= b;
            h *= FnvPrime32;
        }

        return h;
    }

    /// <summary>
    /// The one sampling primitive. Deterministic and stable for the lifetime of
    /// an id.
    /// </summary>
    /// <remarks>
    /// <list type="bullet">
    /// <item>non-finite <paramref name="ratio"/> → IN (fail open — telemetry
    /// going dark on a config hiccup is the failure ADR-097 forbids;
    /// <c>x &lt; NaN</c> is false everywhere, so the naive path would sample
    /// <em>everything</em> out)</item>
    /// <item><c>ratio &lt;= 0.0</c> → OUT, <c>ratio &gt;= 1.0</c> → IN, both
    /// taken before any float math so 1.0 can never drop and 0.0 can never
    /// keep</item>
    /// <item>otherwise <c>(hash / 2^32) &lt; ratio</c>, strict less-than</item>
    /// </list>
    /// </remarks>
    /// <param name="id">Session id, or trace id where a trace exists.</param>
    /// <param name="ratio">0.0 (never) .. 1.0 (always).</param>
    public static bool SampleDecision(string id, double ratio)
    {
        if (double.IsNaN(ratio) || double.IsInfinity(ratio))
        {
            return true;
        }

        if (ratio <= 0.0)
        {
            return false;
        }

        if (ratio >= 1.0)
        {
            return true;
        }

        return Fnv1a32(id) / TwoPow32 < ratio;
    }

    /// <summary>The canonical uppercase spelling written to the wire.</summary>
    public static string ToWire(this CanonicalLevel level) => level switch
    {
        CanonicalLevel.Trace => "TRACE",
        CanonicalLevel.Debug => "DEBUG",
        CanonicalLevel.Info => "INFO",
        CanonicalLevel.Warn => "WARN",
        CanonicalLevel.Error => "ERROR",
        CanonicalLevel.Fatal => "FATAL",
        _ => "INFO",
    };

    // Ordering used by the minimum-level filter. Matches OTel severity numbers.
    private static int Rank(CanonicalLevel level) => level switch
    {
        CanonicalLevel.Trace => 1,
        CanonicalLevel.Debug => 5,
        CanonicalLevel.Info => 9,
        CanonicalLevel.Warn => 13,
        CanonicalLevel.Error => 17,
        CanonicalLevel.Fatal => 21,
        _ => 9,
    };

    /// <summary>
    /// Strict parse: a known level spelling (case-insensitive, surrounding
    /// whitespace trimmed), or null. The alias set is part of the parity
    /// contract — every SDK accepts exactly these spellings.
    /// </summary>
    public static CanonicalLevel? ParseLevel(string level)
    {
        ArgumentNullException.ThrowIfNull(level);

        return level.Trim().ToLowerInvariant() switch
        {
            "trace" or "verbose" => CanonicalLevel.Trace,
            "debug" => CanonicalLevel.Debug,
            "info" or "information" or "log" or "notice" => CanonicalLevel.Info,
            "warn" or "warning" => CanonicalLevel.Warn,
            "error" or "err" => CanonicalLevel.Error,
            "fatal" or "critical" or "crit" or "emergency" or "panic" => CanonicalLevel.Fatal,
            _ => null,
        };
    }

    /// <summary>
    /// Normalize any level spelling to the canonical form. Unknown spellings
    /// become INFO — fail-safe: an unrecognised level must never cause a drop,
    /// and must never be promoted into ERROR (which would corrupt the error
    /// rate).
    /// </summary>
    public static CanonicalLevel NormalizeLevel(string level) => ParseLevel(level) ?? CanonicalLevel.Info;

    /// <summary>True when <paramref name="level"/> is at or above <paramref name="minimum"/>.</summary>
    public static bool MeetsMinimumLevel(CanonicalLevel level, CanonicalLevel minimum) => Rank(level) >= Rank(minimum);

    /// <summary>
    /// The single decision point for "does this log line get emitted?".
    /// </summary>
    /// <remarks>
    /// Order matters and is part of the parity contract:
    /// <list type="number">
    /// <item>kill switch — off means off, no exceptions</item>
    /// <item>minimum level — below the floor is not emitted</item>
    /// <item>WARN/ERROR/FATAL — always 100% (ADR-010: "sampling errors is malpractice")</item>
    /// <item>trace decision, if a trace exists — inherited, never re-rolled</item>
    /// <item>otherwise the session decision — one roll for the whole session</item>
    /// </list>
    /// </remarks>
    public static bool ShouldEmitLog(LogSamplingInput input)
    {
        if (!input.Enabled)
        {
            return false;
        }

        var level = NormalizeLevel(input.Level);
        if (!MeetsMinimumLevel(level, input.MinimumLevel))
        {
            return false;
        }

        if (Rank(level) >= Rank(CanonicalLevel.Warn))
        {
            return true;
        }

        if (input.TraceSampled is { } traceSampled)
        {
            return traceSampled;
        }

        return SampleDecision(input.SessionId, input.LogSamplingRatio);
    }
}
